### ISSUE TITLE
Haven't been able to use patchwork on my desktop for months, it never stops scuttling

### ISSUE THREAD

**Issue by Bortseb** [ssbc/patchwork#1237]
At some point a few months back this started happening after I installed a new patch if I recall correctly.

The following is all I see no matter how many times I try restarting patchwork. I also just installed the latest patch today hoping it might fix it, but no luck.

![image](https://user-images.githubusercontent.com/1617130/72125083-32653c00-3335-11ea-8630-de62c8d803b7.png)

Any clues on what might have gone wrong or how to fix this?

**Comment by thosmos** [comment]
I just copied over my .ssb from an old system and downloaded the new patchwork and I'm getting the same thing.  It seems to run fine with a brand-new .ssb.  I'm wondering what the minimum that is needed to keep my identity but enable the app to load?

**Comment by thosmos** [comment]
Have you taken a look at the log?  I had a similar problem that was easy to fix once I looked at the logs (I was running a service on port 8989 which conflicted with the app which needs ports 8989 and 8008 by default)

**Comment by black-puppydog** [comment]
Hey @Bortseb, sorry to hear you're having trouble with patchwork.
Without any of the logs, I'd imagine it got stuck indexing somewhere, simply because for me that's the most common cause for this kind of behaviour, but it could be something else too.
I'll echo a [comment](https://ssb.muchmuch.coffee/%25IRfuu6iADMFm7yvU4%2FYFX56d1E%2BLbDcagifzX9TBk4g%3D.sha256) from @mixmix on this:

> I would honestly try closing patchwork, installing #patchbay and opening that - for some reason it can un-stuck #patchwork sometimes… I do not know why. Then if that fixes it close patchbay and open patchwork again @kel if you want to continue with that.
> 
> If none of that works, step up to nuking the indexes (the quick lookup tables). These can be safely regrown from the core log. If you want to try this get in touch. You can delete basically everything in ~/.ssb/flume EXCEPT ~/.ssb/flume/log.offset (that there is your spinal chord, don’t touch it).
> 
> If none of that works, then get on a call with someone.

Of course, if you can launch patchwork from the command line, that would make for easier debugging too. Simply grab the latest AppImage, make it executable (via UI or with `chmod +x <filename>`) and then run it (probably something like `~/Downloads/<patchwork_file>.AppImage`)

**Comment by christianbundy** [comment]
Really sorry that you're having this problem, that sucks. Could you try starting Patchwork from the command-line and paste the relevant output? I'm guessing that there's an error happening somewhere, and we need to fix it.

**Comment by black-puppydog** [comment]
As it happens, @Powersource has been working on a tool for just such an occasion:

https://github.com/ssbc/ssb-first-aid-kit/releases/tag/v1.3.0

It might be a good time to try it out? In any case, please make a backup of your identity (`~/.ssb/secret`) before trying anything.

**Comment by christianbundy** [comment]
Does that tool tell us what the error is? It's important to me that we know what the problem is before we start deleting things, if possible.

**Comment by Powersource** [comment]
If the indexes are stuck, it *should* be pretty clear in that app. Haven't
gotten them stuck for myself to try tho.

Jacob

**Comment by Bortseb** [comment]
> Really sorry that you're having this problem, that sucks. Could you try starting Patchwork from the command-line and paste the relevant output? I'm guessing that there's an error happening somewhere, and we need to fix it.

How do I do this?



> As it happens, @Powersource has been working on a tool for just such an occasion:
> 
> https://github.com/ssbc/ssb-first-aid-kit/releases/tag/v1.3.0
> 
> It might be a good time to try it out? In any case, please make a backup of your identity (`~/.ssb/secret`) before trying anything.

I tried this, but it doesn't seem to do anything for me:
![image](https://user-images.githubusercontent.com/1617130/72670569-bc379800-3a0c-11ea-8907-8953df54ce3b.png)

**Comment by christianbundy** [comment]
> How do I do this?

Oh, just realized you're on Windows (I think). I'm not sure that there's an easy way to do this.

Could you try "delete indexes" when Patchwork is stopped, and then restart Patchwork?

**Comment by Bortseb** [comment]
yea, im on windows... ill try that now.

**Comment by Bortseb** [comment]
yea meant delete indexes in the first aid app right? that didn't work

**Comment by Powersource** [comment]
> I'm not sure that there's an easy way to do this.

@christianbundy Is this as in "can't really do it (this way) because that's not how windows works" or "you can do it but it's pretty convoluted" (was a while since i used windows so idk :P ) If it's just hard we could maybe ask some windows dev to help here.

**Comment by christianbundy** [comment]
@Powersource 

I mean, I think we just want them to open `patchwork.exe` from the command-line so that we get error output, but I don't remember how Windows `$PATH`-type stuff works well enough to give someone a command that they can copy and paste. I'm not sure about capitalization or if it has an `ssb-` prefix.

**Comment by Bortseb** [comment]
I'm not at home for a week so can't test. But I will run whatever command you suggest to see error logs etc... I just don't know what command to run.

I have git bash installed on my system, so maybe a command you are more familiar with would work?

I can fiddle with path variables too if things don't initially seem to want to work... I just don't know what ssb command to run.

**Comment by christianbundy** [comment]
I think you should be running `patchwork.exe` or `ssb-patchwork` or something like that. The problem is that I'm not sure how you run Patchwork from the CLI on Windows. I think we'll need to either have you check a few variations or wait for someone who knows the command.

**Comment by Powersource** [comment]
cc windows squad @soapdog @cblgh (type *unsubscribe* if you want me to stop summoning you :stuck_out_tongue:) do you know how to run that command? :)

**Comment by cblgh** [comment]
_unsubscribe_

**Comment by soapdog** [comment]
*A wild SoapDog appears on the thread*

Ok? What's the objective here? Running Patchwork from the command line?

## Running `patchwork.exe` from the command-line

Open `powershell` which is on the _Start Menu_ under _Windows Powershell_ entry. You'll end up in a nice blue screen terminal:

![image](https://user-images.githubusercontent.com/23247/72889963-c4435080-3d08-11ea-927a-b4a782d93e8f.png)

If you've installed Patchwork using the `Patchwork-Setup.exe`, it will be installed to the default location. In my own case it is `C:\Users\andre\AppData\Local\Programs\ssb-patchwork\Patchwork.exe` be aware that the user name in that case will change, yours won't be `andre`. Lucky for you, powershell already opens in the users folder, so to go into the folder where Patchwork is installed you just need to type:

````
cd AppData\Local\Programs\ssb-patchwork
````

After that you can type `dir` to check which files are present and make sure you're in the correct place:

![image](https://user-images.githubusercontent.com/23247/72890267-6ebb7380-3d09-11ea-82e9-dcb4244a4538.png)

If you're in the right place, you can just type:

````
.\Patchwork.exe
````
To run it, there will be a ton of output from it on the powershell.

**Comment by Bortseb** [comment]
Thanks! just got home so I can try this now.

I ran the command, the beginning starts with a what looks like a JSON object with my configurations etc... if thats safe to share, let me know and I'll paste it too... But there error I see at the bottom is this

```
git-ssb is not installed, or not available in path
ssb-friends: stream legacy api used

C:\Users\Rob\AppData\Local\Programs\ssb-patchwork\resources\app.asar\node_modules\levelup\lib\levelup.js:119
      return callback(new OpenError(err))
                      ^
Error [OpenError]: IO error: C:\Users\Rob\.ssb\blobs_push/MANIFEST-000582: The system cannot find the file specified.

    at C:\Users\Rob\AppData\Local\Programs\ssb-patchwork\resources\app.asar\node_modules\levelup\lib\levelup.js:119:23
    at C:\Users\Rob\AppData\Local\Programs\ssb-patchwork\resources\app.asar\node_modules\deferred-leveldown\node_modules\abstract-leveldown\abstract-leveldown.js:36:14
    at C:\Users\Rob\AppData\Local\Programs\ssb-patchwork\resources\app.asar\node_modules\deferred-leveldown\deferred-leveldown.js:31:21
    at C:\Users\Rob\AppData\Local\Programs\ssb-patchwork\resources\app.asar\node_modules\encoding-down\node_modules\abstract-leveldown\abstract-leveldown.js:36:14
    at C:\Users\Rob\AppData\Local\Programs\ssb-patchwork\resources\app.asar\node_modules\leveldown\node_modules\abstract-leveldown\abstract-leveldown.js:30:14
Emitted 'error' event at:
    at C:\Users\Rob\AppData\Local\Programs\ssb-patchwork\resources\app.asar\node_modules\levelup\lib\levelup.js:60:19
    at C:\Users\Rob\AppData\Local\Programs\ssb-patchwork\resources\app.asar\node_modules\levelup\lib\levelup.js:119:14
    at C:\Users\Rob\AppData\Local\Programs\ssb-patchwork\resources\app.asar\node_modules\deferred-leveldown\node_modules\abstract-leveldown\abstract-leveldown.js:36:14
    [... lines matching original stack trace ...]
    at C:\Users\Rob\AppData\Local\Programs\ssb-patchwork\resources\app.asar\node_modules\leveldown\node_modules\abstract-leveldown\abstract-leveldown.js:30:14
```

**Comment by Powersource** [comment]
The config should be safe to share, the private key is hidden in that printout, but I don't think the config would help here.

Found an old message by Mikey saying

> the blobs_push folder is a leveldb used by ssb-blobs to keep track of which blobs you are “pushing”

If you've backed stuff up I would recommend just deleting the entire `C:\Users\Rob\.ssb\blobs_push/` folder.

**Comment by Bortseb** [comment]
I just tried deleting the blobs_push folder and tested, this is what I get back after it still gets stuck at scuttling...

```
PS C:\Users\Rob\AppData\Local\Programs\ssb-patchwork> .\Patchwork.exe
PS C:\Users\Rob\AppData\Local\Programs\ssb-patchwork>
 {
  path: 'C:\\Users\\Rob\\.ssb',
  party: true,
  timeout: 0,
  pub: true,
  local: true,
  friends: { dunbar: 150, hops: 2 },
  gossip: { connections: 3, autoPopulate: false },
  connections: {
    outgoing: {
      net: [ { transform: 'shs' } ],
      onion: [ { transform: 'shs' } ],
      tunnel: [ { transform: 'shs' } ]
    },
    incoming: {
      net: [
        {
          host: '10.161.9.192',
          port: 8008,
          scope: [ 'local' ],
          transform: 'shs'
        },
        {
          host: '::1',
          port: 8008,
          scope: [ 'device' ],
          transform: 'shs'
        },
        {
          host: '127.0.0.1',
          port: 8008,
          scope: [ 'device' ],
          transform: 'shs'
        },
        {
          host: '10.0.75.1',
          port: 8008,
          scope: [ 'local' ],
          transform: 'shs'
        },
        {
          host: '192.168.0.10',
          port: 8008,
          scope: [ 'local' ],
          transform: 'shs'
        },
        {
          host: '172.17.73.1',
          port: 8008,
          scope: [ 'local' ],
          transform: 'shs'
        }
      ],
      ws: [
        {
          host: '10.161.9.192',
          port: 8989,
          scope: [ 'local' ],
          transform: 'shs'
        },
        {
          host: '::1',
          port: 8989,
          scope: [ 'device' ],
          transform: 'shs'
        },
        {
          host: '127.0.0.1',
          port: 8989,
          scope: [ 'device' ],
          transform: 'shs'
        },
        {
          host: '10.0.75.1',
          port: 8989,
          scope: [ 'local' ],
          transform: 'shs'
        },
        {
          host: '192.168.0.10',
          port: 8989,
          scope: [ 'local' ],
          transform: 'shs'
        },
        {
          host: '172.17.73.1',
          port: 8989,
          scope: [ 'local' ],
          transform: 'shs'
        }
      ],
      tunnel: [ { scope: 'public', transform: 'shs' } ]
    }
  },
  timers: { connection: 0, reconnect: 5000, ping: 300000, handshake: 5000 },
  caps: {
    shs: '1KHLiKZvAvjbY1ziZEHMXawbCEIM6qwjCDm3VYRan/s=',
    sign: null,
    invite: 'HT0wIYuk3OWc2FtaCfHNnakV68jSGRrjRMP9Kos7IQc='
  },
  master: [],
  logging: { level: 'notice' },
  port: 8008,
  blobsPort: 8989,
  server: true,
  _: [],
  host: '10.161.9.192',
  ws: {
    host: '10.161.9.192',
    port: 8989,
    scope: [ 'local' ],
    transform: 'shs'
  },
  keys: {
    curve: 'ed25519',
    public: 'Zh2Smed6SMinqugfEc/7lQlziJmRZNm+jzQFxmbNAjQ=.ed25519',
    private: null,
    id: '@Zh2Smed6SMinqugfEc/7lQlziJmRZNm+jzQFxmbNAjQ=.ed25519'
  },
  remote: 'net:127.0.0.1:8008~shs:Zh2Smed6SMinqugfEc/7lQlziJmRZNm+jzQFxmbNAjQ='
}
git-ssb is not installed, or not available in path
ssb-friends: stream legacy api used

C:\Users\Rob\AppData\Local\Programs\ssb-patchwork\resources\app.asar\node_modules\levelup\lib\levelup.js:119
      return callback(new OpenError(err))
                      ^
Error [OpenError]: IO error: C:\Users\Rob\.ssb\db/MANIFEST-000582: The system cannot find the file specified.

    at C:\Users\Rob\AppData\Local\Programs\ssb-patchwork\resources\app.asar\node_modules\levelup\lib\levelup.js:119:23
    at C:\Users\Rob\AppData\Local\Programs\ssb-patchwork\resources\app.asar\node_modules\deferred-leveldown\node_modules\abstract-leveldown\abstract-leveldown.js:36:14
    at C:\Users\Rob\AppData\Local\Programs\ssb-patchwork\resources\app.asar\node_modules\deferred-leveldown\deferred-leveldown.js:31:21
    at C:\Users\Rob\AppData\Local\Programs\ssb-patchwork\resources\app.asar\node_modules\encoding-down\node_modules\abstract-leveldown\abstract-leveldown.js:36:14
    at C:\Users\Rob\AppData\Local\Programs\ssb-patchwork\resources\app.asar\node_modules\leveldown\node_modules\abstract-leveldown\abstract-leveldown.js:30:14
Emitted 'error' event at:
    at C:\Users\Rob\AppData\Local\Programs\ssb-patchwork\resources\app.asar\node_modules\levelup\lib\levelup.js:60:19
    at C:\Users\Rob\AppData\Local\Programs\ssb-patchwork\resources\app.asar\node_modules\levelup\lib\levelup.js:119:14
    at C:\Users\Rob\AppData\Local\Programs\ssb-patchwork\resources\app.asar\node_modules\deferred-leveldown\node_modules\abstract-leveldown\abstract-leveldown.js:36:14
    [... lines matching original stack trace ...]
    at C:\Users\Rob\AppData\Local\Programs\ssb-patchwork\resources\app.asar\node_modules\leveldown\node_modules\abstract-leveldown\abstract-leveldown.js:30:14
```

After that didn't work... I stepped it up the also take the advice above

> If none of that works, step up to nuking the indexes (the quick lookup tables). These can be safely regrown from the core log. If you want to try this get in touch. You can delete basically everything in ~/.ssb/flume EXCEPT ~/.ssb/flume/log.offset (that there is your spinal chord, don’t touch it).

That also didn't work, and gave the same log as I just pasted above.

I notice that the error is looking for a MANIFEST-000582

before deleting my blobs_push folder I made a copy of everything and there wasn't a manifest file in there.. I notice when I try and launch patchwork it creates a manifest file, but not matching this one its looking for, its instead MANIFEST-000002

I also didn't realize how big my .ssb folder was until today :P like 9GB

**Comment by christianbundy** [comment]
> That also didn't work, and gave the same log as I just pasted above.

If that didn't work, I'd delete absolutely everything in `~/.ssb` except for:

- `secret`
- `gossip.json`
- `conn.json`
- `flume/log.offset`

Everything else should be generated from scratch. If *that* doesn't work, then I'm not sure what else we can even try. I've never seen that error before, and I feel bad that you have to spend so much time debugging this and trying to get back on the network. Good luck! :crossed_fingers:

**Comment by Bortseb** [comment]
ok thanks everyone here for your help... this last peice of advice from @christianbundy finally fixed it for me! :)

Have I lost anything by deleting everything? Or now that this is done indexing/updating I am back to where I was before? i.e., is it safe to now delete the backup copy I made of my old ssb folder?

**Comment by Bortseb** [comment]
Well, I must have lost something as my old folder was 9GB, and my new one is only 1.84GB... Is there any intelligent way to copy files from the old one into the new one? Or that would just make more of a mess?

**Comment by christianbundy** [comment]
My guess is that you're missing your old blobs -- when the new `~/.ssb/` sees a blob it needs it'll download it and place it in `~/.ssb/blobs/sha256/`, and since you've only been using Patchwork for a few minutes it probably hasn't downloaded many of those. You *should* be able to copy and paste blobs from the old directory into the new one, but you'll need to make sure it keeps the same directory structure.

**Comment by Bortseb** [comment]
I just copied over all the blobs, but skipped replacing any blobs with the same name that was created after my most recent successful boot of patchwork.

Everything is still working fine after doing this so I am happy now :)

I guess I can close this issue now. Thanks for all your help!

**Comment by christianbundy** [comment]
Thanks for being so patient and fixing it up! 

> I just copied over all the blobs, but skipped replacing any blobs with the same name that was created after my most recent successful boot of patchwork.

If you have a script for this handy, I'm sure others would love to use it. If not, no problem. Thanks again!

**Comment by Bortseb** [comment]
Sorry, no script for it. Windows just does that by default when copying a
bunch of files to a new location, and some of which have the same name.

**Comment by christianbundy** [comment]
Oh, awesome. I was impressed by your specificity and assumed you wrote a script. :grimacing:
